titus-common/src/main/java/com/netflix/titus/common/util/loadshedding/tokenbucket/TokenBucketAdmissionController.java

```diff
@@ -0,0 +1,194 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.titus.common.util.loadshedding.tokenbucket;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+import com.netflix.titus.common.runtime.TitusRuntime;
+import com.netflix.titus.common.util.cache.Cache;
+import com.netflix.titus.common.util.cache.Caches;
+import com.netflix.titus.common.util.limiter.Limiters;
+import com.netflix.titus.common.util.limiter.tokenbucket.TokenBucket;
+import com.netflix.titus.common.util.loadshedding.AdmissionController;
+import com.netflix.titus.common.util.loadshedding.AdmissionControllerRequest;
+import com.netflix.titus.common.util.loadshedding.AdmissionControllerResponse;
+import com.netflix.titus.common.util.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Admission controller with multiple token buckets. A token bucket is selected by evaluating matching criteria in
+ * declaration order until first match is found. The selected bucket is tried, and either success or failure is
+ * returned to the caller.
+ * <p/>
+ * <h1>Example: shared caller bucket</h1>
+ * slowMethods.sharedByCallers=true<br/>
+ * slowMethods.callerPattern=.*<br/>
+ * slowMethods.endpointPattern=create.*<br/>
+ * ...</br>
+ * <p/>
+ * fastMethods.sharedByCallers=true<br/>
+ * fastMethods.callerPattern=.*<br/>
+ * fastMethods.endpointPattern=get.*<br/>
+ * ...</br>
+ * <p/>
+ * Caller Alice and Bob making a call to createJob method will share a single bucket with id (.*, create.*), and
+ * a single bucket for methods getJob/getTask/etc with id (.*, get.*).
+ * <p/>
+ * <h1>Example: per caller bucket</h1>
+ * slowMethods.sharedByCallers=false<br/>
+ * slowMethods.callerPattern=.*<br/>
+ * slowMethods.endpointPattern=create.*<br/>
+ * ...</br>
+ * <p/>
+ * fastMethods.sharedByCallers=false<br/>
+ * fastMethods.callerPattern=.*<br/>
+ * fastMethods.endpointPattern=get.*<br/>
+ * ...</br>
+ * <p/>
+ * Caller Alice and Bob making a call to createJob method will have own buckets with ids (Alice, create.*) and
+ * (Bob, create.*). Similarly, for methods getJob/getTask/etc there will be two buckets (Alice, get.*) and
+ * (Bob, get.*) respectively.
+ */
+public class TokenBucketAdmissionController implements AdmissionController {
+
+    private static final Logger logger = LoggerFactory.getLogger(TokenBucketAdmissionController.class);
+
+    private static final AdmissionControllerResponse DEFAULT_OK = AdmissionControllerResponse.newBuilder()
+            .withAllowed(true)
+            .withReasonMessage("Rate limits not configured")
+            .build();
+
+    private static final int MAX_CACHE_SIZE = 10_000;
+    private static final Duration CACHE_ITEM_TIMEOUT = Duration.ofSeconds(60);
```

Comment:
is this timeout needed considering there is a max cache size?

Reply:
Bumping up timeout to 5min. If the cache gets full, some items will be removed. From caffeine documentation:
```   
/* Specifies the maximum number of entries the cache may contain. Note that the cache <b>may evict
   * an entry before this limit is exceeded or temporarily exceed the threshold while evicting</b>.
   * As the cache size grows close to the maximum, the cache evicts entries that are less likely to
   * be used again. For example, the cache may evict an entry because it hasn't been used recently
   * or very often.
 */
public Caffeine<K, V> maximumSize(@Nonnegative long maximumSize) {
```